Threads don't get removed from cache if the parent channel is deleted, causing AssertionErrors

When you create a thread (tested with a public one, most likely affects private threads too) in a channel and then later delete said channel, the thread is not removed from Javacord's channel cache. This causes the channel cache to contain threads without a parent channel (since the deleted parent correctly gets removed from cache) - which again introduces the dreaded `AssertionError: Thread has no parent channel` when trying to access that server's channel list through methods such as `Server.getChannels()` :D

I think this is because there is no special handling for threads in the `ChannelDeleteHandler`, and Discord doesn't send `THREAD_DELETE` packets if the parent channel is deleted. As mentioned in the Discord documentation:

> When a client loses access to a channel, the gateway does not send them a `Thread Delete` event or any equivalent. They will still receive _an_ event when this happens, it just will not be a thread-specific event.

I have not tested whether this also leaves behind stray ThreadChannel instances in other caches such as the global channel cache, and whether that also happens when you delete the server as opposed to just the channel. But those are things that came to my mind which have to be looked out for.

This was tested using a mostly up-to-date Javacord version at f0171112064a48b60ff90bdb92a326f112366274.

In the same go, I suggest exposing the internal function `getUnorderedChannels()`, for cases where you want to access the channel list but don't care about it being sorted, to save some performance by not needlessly sorting the list.

Demonstration:

https://user-images.githubusercontent.com/10065021/174498712-0e77a29b-a059-4c33-830e-f812cfd56cfd.mp4

--- a/javacord-api/src/main/java/org/javacord/api/entity/server/Server.java
+++ b/javacord-api/src/main/java/org/javacord/api/entity/server/Server.java
@@ -2392,6 +2392,13 @@ public interface Server extends DiscordEntity, Nameable, UpdatableFromCache<Serv
     List<ServerChannel> getChannels();
 
     /**
+     * Gets an unordered collection with all channels in the server.
+     *
+     * @return An unordered collection with all channels in the server.
+     */
+    Set<ServerChannel> getUnorderedChannels();
+
+    /**
      * Gets a sorted list (by position) with all regular channels of the server.
      *
      * @return A sorted list (by position) with all regular channels of the server.

--- a/javacord-core/src/main/java/org/javacord/core/DiscordApiImpl.java
+++ b/javacord-core/src/main/java/org/javacord/core/DiscordApiImpl.java
@@ -12,6 +12,7 @@ import org.apache.logging.log4j.Logger;
 import org.javacord.api.DiscordApi;
 import org.javacord.api.Javacord;
 import org.javacord.api.entity.ApplicationInfo;
+import org.javacord.api.entity.DiscordEntity;
 import org.javacord.api.entity.activity.Activity;
 import org.javacord.api.entity.activity.ActivityType;
 import org.javacord.api.entity.channel.Channel;
@@ -959,6 +960,19 @@ public class DiscordApiImpl implements DiscordApi, DispatchQueueSelector {
             if (channel == null) {
                 return cache;
             }
+
+            //Remove all ServerThreadChannels when the parent channel is removed
+            channel.asServerChannel().ifPresent(serverChannel -> {
+                if (serverChannel.asServerThreadChannel().isPresent()) {
+                    return;
+                }
+
+                serverChannel.getServer().getThreadChannels().stream()
+                        .filter(c -> c.getParent().getId() == serverChannel.getId())
+                        .mapToLong(DiscordEntity::getId)
+                        .forEach(this::removeChannelFromCache);
+            });
+
             if (channel instanceof Cleanupable) {
                 ((Cleanupable) channel).cleanup();
             }

--- a/javacord-core/src/main/java/org/javacord/core/entity/channel/ServerThreadChannelImpl.java
+++ b/javacord-core/src/main/java/org/javacord/core/entity/channel/ServerThreadChannelImpl.java
@@ -8,6 +8,7 @@ import org.javacord.api.util.cache.MessageCache;
 import org.javacord.core.DiscordApiImpl;
 import org.javacord.core.entity.server.ServerImpl;
 import org.javacord.core.listener.channel.server.text.InternalServerTextChannelAttachableListenerManager;
+import org.javacord.core.util.Cleanupable;
 import org.javacord.core.util.cache.MessageCacheImpl;
 import org.javacord.core.util.rest.RestEndpoint;
 import org.javacord.core.util.rest.RestMethod;
@@ -22,8 +23,8 @@ import java.util.concurrent.CompletableFuture;
 /**
  * The implementation of {@link ServerThreadChannel}.
  */
-public class ServerThreadChannelImpl extends ServerChannelImpl implements ServerThreadChannel, InternalTextChannel,
-        InternalServerTextChannelAttachableListenerManager {
+public class ServerThreadChannelImpl extends ServerChannelImpl implements ServerThreadChannel, Cleanupable,
+        InternalTextChannel, InternalServerTextChannelAttachableListenerManager {
 
     /**
      * The message cache of the server text channel.
@@ -213,5 +214,15 @@ public class ServerThreadChannelImpl extends ServerChannelImpl implements Server
     @Override
     public String getMentionTag() {
         return "<#" + getIdAsString() + ">";
+    }
+
+    @Override
+    public void cleanup() {
+        messageCache.cleanup();
+
+        ((DiscordApiImpl) getApi()).forEachCachedMessageWhere(
+                msg -> msg.getChannel().getId() == getId(),
+                msg -> ((DiscordApiImpl) getApi()).removeMessageFromCache(msg.getId())
+        );
     }
 }

--- a/javacord-core/src/main/java/org/javacord/core/entity/server/ServerImpl.java
+++ b/javacord-core/src/main/java/org/javacord/core/entity/server/ServerImpl.java
@@ -1066,12 +1066,8 @@ public class ServerImpl implements Server, Cleanupable, InternalServerAttachable
         this.serverFeatures.addAll(serverFeatures);
     }
 
-    /**
-     * Gets an unordered collection with all channels in the server.
-     *
-     * @return An unordered collection with all channels in the server.
-     */
-    public Collection<ServerChannel> getUnorderedChannels() {
+    @Override
+    public Set<ServerChannel> getUnorderedChannels() {
         return api.getEntityCache().get().getChannelCache().getChannelsOfServer(getId());
     }
 

--- a/javacord-core/src/main/java/org/javacord/core/util/handler/channel/ChannelDeleteHandler.java
+++ b/javacord-core/src/main/java/org/javacord/core/util/handler/channel/ChannelDeleteHandler.java
@@ -168,13 +168,7 @@ public class ChannelDeleteHandler extends PacketHandler {
         long channelId = channelJson.get("id").asLong();
         api.getPossiblyUnreadyServerById(serverId)
                 .flatMap(server -> server.getForumChannelById(channelId))
-                .ifPresent(channel -> {
-                    dispatchServerChannelDeleteEvent(channel);
-                    api.forEachCachedMessageWhere(
-                            msg -> msg.getChannel().getId() == channelId,
-                            msg -> api.removeMessageFromCache(msg.getId())
-                    );
-                });
+                .ifPresent(this::dispatchServerChannelDeleteEvent);
         api.removeObjectListeners(ServerForumChannel.class, channelId);
         api.removeObjectListeners(RegularServerChannel.class, channelId);
         api.removeObjectListeners(ServerChannel.class, channelId);

--- a/javacord-core/src/main/java/org/javacord/core/util/handler/channel/thread/ThreadDeleteHandler.java
+++ b/javacord-core/src/main/java/org/javacord/core/util/handler/channel/thread/ThreadDeleteHandler.java
@@ -51,19 +51,13 @@ public class ThreadDeleteHandler extends PacketHandler {
         final long channelId = packet.get("id").asLong();
         api.getPossiblyUnreadyServerById(serverId)
                 .flatMap(server -> server.getThreadChannelById(channelId))
-                .ifPresent(channel -> {
-                    dispatchThreadDeleteEvent(channel);
-                    api.forEachCachedMessageWhere(
-                            msg -> msg.getChannel().getId() == channelId,
-                            msg -> api.removeMessageFromCache(msg.getId())
-                    );
-                });
+                .ifPresent(this::dispatchThreadDeleteEvent);
         api.removeObjectListeners(ServerThreadChannel.class, channelId);
         api.removeObjectListeners(ServerChannel.class, channelId);
         api.removeObjectListeners(TextChannel.class, channelId);
         api.removeObjectListeners(Channel.class, channelId);
 
-        api.removeChannelFromCache(packet.get("id").asLong());
+        api.removeChannelFromCache(channelId);
     }
 
     /**

